### PR TITLE
Refactor Projector based on latest widgets

### LINKED ISSUE #129: Widget Projector
**Enhancement** 

With the introduction of `d` and the change within `widgets` generally, it would be beneficial to reimplement `projector` using `createWidgetBase` as the base and providing familiar style API to create projectors for a consumers application.

Something alone the lines of:

```ts
const children = (this: Projector): DNode[] => {
    const appOptions: AppOptions = {
        id: 'app',
        stateFrom: widgetStore
    };

    return [ d(createApp, appOptions) ];
});

const projector = createProjector({ root, children });

projector.attach().then(() {
    console.log('Im attached');
});
```

Package Version: latest

### PR DESCRIPTION
**Pull request by agubler**
**Type:** feature

The following has been addressed in the PR:

* [x] There is a related issue
* [x] All code matches the [style guide](https://github.com/dojo/meta/blob/master/STYLE.md)
* [x] Unit or Functional tests are included in the PR

**Description:**

Reimplement `projector` inline with `createWidgetBase` and `d`

Resolves #129

### CODE REVIEW

**Comment by kitsonk** [comment]
Why no default singleton anymore?  That is likely to be a serious breaking change.

**Comment by agubler** [comment]
@kitsonk certainly we can add that, my preference is not to support a default singleton inline with all other widgets. It is a breaking change, but fortunately we only break ourselves (examples) which we can fix prior to moving to the version of widgets that contains these changes.

**Comment by Tomdye** [comment]
We're going to have to fix all the examples before release anyway as per the other changes

**Comment by agubler** [comment]
@kitsonk although that does then put the responsibility on the consumer to know when to use this projector and when to use the "node" impl of the projector.....

**Comment by kitsonk** [comment]
> default singleton inline with all other widgets

A projector is a different kettle of fish... you need a projector to display anything...  having a default singleton that out binds to the root of the page is far more usable...  breaking out the factory is fine, but there should still be a singleton module so people can `import projector from 'dojo-widgets/projector'` and be done with it.

**Comment by agubler** [comment]
@kitsonk okay, that seems reasonable :) I'll work towards that.

**Comment by agubler** [comment]
Having thought about it, a singleton actually doesn't make any sense now and would not actually be usable. As in you need to create the projector with widgets attached via options or extend the projector and override the getChildrenNodes function.

A projector is now treated and consumed as any widget.

**Comment by Tomdye** [comment]
@agubler it might be worth adding an example of intended usage to this PR for clarity

File: src/createProjector.ts [hunk 1]
```diff
@@ -0,0 +1,171 @@
...
+
+/**
+ * Private state map keyed by instance.
+ */
+const projectorDataMap = new WeakMap<Projector, ProjectorData>();
+
+/**
+ * Projector Factory
+ */
+const createProjector: ProjectorFactory = cretateWidgetBase
```
Comment (Tomdye) [inline src/createProjector.ts]: you've spelt `createWidgetBase` wrong.

File: src/createProjector.ts [hunk 2]
```diff
@@ -0,0 +1,171 @@
+import { ComposeFactory } from 'dojo-compose/compose';
+import { Handle } from 'dojo-interfaces/core';
+import { VNodeProperties } from 'dojo-interfaces/vdom';
+import { Widget, WidgetState, WidgetOptions } from 'dojo-interfaces/widgetBases';
+import WeakMap from 'dojo-shim/WeakMap';
+import { createProjector as createMaquetteProjector, Projector as MaquetteProjector } from 'maquette';
+import cretateWidgetBase from './bases/createWidgetBase';
```
Comment (Tomdye) [inline src/createProjector.ts]: spelling

**Comment by agubler** [comment]
@Tomdye I'll add both documentation and update the examples once I have completed the tests

File: src/createProjector.ts (hunk 1 above)
Comment (agubler) [inline src/createProjector.ts]: Ha. It works as I spelt it equally incorrectly in the import.

**Comment by agubler** [comment]
@kitsonk this is ready to go now with tests and updated example.

Doesn't export a singleton as it would not be useful for consumers with this implementation.

**Comment by Tomdye** [comment]
Can you also please remove `mixins/interfaces`. It it only used within the `projector.ts` file you have deleted.

**Comment by agubler** [comment]
It's used in `src/util/lang.ts`, which still exists in my branch.

**Comment by Tomdye** [comment]
OK fair enough for the moment. Once the second tidy up PR has been merged (https://github.com/dojo/widgets/pull/136) `src/util/lang` will no longer exist.

**Comment by Tomdye** [comment]
Should be able to just delete

File: src/createProjector.ts [hunk 3]
```diff
@@ -0,0 +1,187 @@
...
+	.mixin({
+		mixin: {
+			nodeAttributes: [
+				function(this: Projector): VNodeProperties {
+					const { afterCreate } = projectorDataMap.get(this);
+					return { afterCreate };
+				}
+			]
+		},
+		aspectAdvice: {
```
Comment (kitsonk) [inline src/createProjector.ts]: What aspect advice instead of just listening for `invalidated` event?

File: src/createProjector.ts [hunk 4]
```diff
@@ -0,0 +1,187 @@
...
+			get root(this: Projector): Element {
+				const projectorData = projectorDataMap.get(this);
+				return projectorData && projectorData.root;
+			},
+
+			get projector(this: Projector): MaquetteProjector {
+				return projectorDataMap.get(this).projector;
+			},
+
+			get state(this: Projector): ProjectorState {
```
Comment (kitsonk) [inline src/createProjector.ts]: Why are you overriding this?  Why does `Stateful` not meet this use case?

File: src/createProjector.ts [hunk 5]
```diff
@@ -0,0 +1,187 @@
...
+				}
+			]
+		},
+		aspectAdvice: {
+			after: {
+				invalidate(this: Projector): void {
+					const projectorData = projectorDataMap.get(this);
+					if (projectorData.state === ProjectorState.Attached) {
+						this.emit({
+							type: 'schedulerender',
```
Comment (kitsonk) [inline src/createProjector.ts]: probably should be `render:scheduled` to align to other event type names

File: src/createProjector.ts [hunk 6]
```diff
@@ -0,0 +1,187 @@
...
+							projectorData.projector.detach(render);
+							projectorData.state = ProjectorState.Detached;
+						}
+						projectorData.attachHandle = { destroy() {} };
+					}
+				});
+
+				projectorData.attachPromise = new Promise((resolve, reject) => {
+					projectorData.afterCreate = () => {
+						this.emit({ type: 'attach' });
```
Comment (kitsonk) [inline src/createProjector.ts]: Should be named `attached` and have a target.

File: src/createProjector.ts [hunk 7]
```diff
@@ -0,0 +1,187 @@
...
+interface ProjectorData {
+	projector: MaquetteProjector;
+	root: Element;
+	state: ProjectorState;
+	attachPromise?: Promise<Handle>;
+	attachHandle?: Handle;
+	afterCreate?: () => void;
+}
+
+export type Projector = Widget<WidgetState> & ProjectorMixin;
```
Comment (kitsonk) [inline src/createProjector.ts]: Why should have the overloads for `on()` with the events it emits...

**Comment by kitsonk** [comment]
I am not going to stand in the way of this, though I still think it is significantly less usable not having any sort of API to attach widgets to a projector post creation and to have a singleton that attaches to the the `document.body`.  I might be wrong...

Also, I would consider how we could attach children via the state by giving widget IDs in `children`.  That could obviously be added later though.

**Comment by agubler** [comment]
> Also, I would consider how we could attach children via the state by giving widget IDs in children. That could obviously be added later though.

This would be supported when we finalise the "container" support that will go in to `createWidgetBase`.

File: src/createProjector.ts (hunk 6 above)
Comment (agubler) [inline src/createProjector.ts]: what about `projector:attached`?

File: src/createProjector.ts (hunk 3 above)
Comment (agubler) [inline src/createProjector.ts]: can update to listen to the `invalidated` event

File: src/createProjector.ts (hunk 7 above)
Comment (agubler) [inline src/createProjector.ts]: That would be good, but I cannot get this to work 😢

```ts
export interface ProjectorOverloads {
    on(type: 'render:scheduled', listener: EventListener<Projector, EventTargettedObject<Projector>>);
    on(type: 'projector:attached', listener: EventListener<Projector, EventTargettedObject<Projector>>);
}
```

and added to `export type Projector = Widget<WidgetState> & ProjectorMixin & ProjectorOverloads;` but got an error that included `Types of property 'mixin' are incompatible.`. 

not sure what I am doing wrong.

Comment (agubler) [inline src/createProjector.ts]: It's expecting me to implement `on` in the `mixin` block

**Comment by agubler** [comment]
@kitsonk nudge, i did have an issue trying to add an overloads interface :(

File: src/createProjector.ts (hunk 7 above)
Comment (kitsonk) [inline src/createProjector.ts]: What I think you have to do is cast the `createProjector = createWidgetBase.mixin(...) as ProjectorFactory` and of doing a left hand assignment, but I am not 100% sure

Comment (agubler) [inline src/createProjector.ts]: it didn't seem to work, agree that it would be nice to have - i'll investigate after this is in.

**Pull request merged**